torch/csrc/autograd/engine.cpp

```diff
@@ -421,6 +443,18 @@ auto Engine::evaluate_function(FunctionTask& task) -> void {
           continue;
         }
       }
+
+      #ifdef USE_CUDA
+        // Switches to and syncs with child's stream (if needed)
+        const auto child_device = output.is_cuda() ? output.get_device() : -1;
+        auto child_stream = next.function->input_metadata(next.input_nr).stream();
+        auto_gpu_stream.set(child_device, child_stream);
+        if (parent_device != -1 && child_device != -1 && parent_stream != child_stream) {
+          cuda_event.record(parent_stream);
```

Comment:
This will record the same event multiple times on `parent_stream`. Are we sure this is really what we want? Is it ok for this to happen multiple times?

Reply:
The CUDAEvent checks if it was recorded to and, if so, does nothing. You can think of this as "if (!recorded) record()."

Reply:
Uh this is not obvious at all, and since it seems like a general purpose struct, I'd rather avoid doing this implicitly. Maybe we could have an `IdempotentCUDAEvent`, or any other name that would signify it's a one-time thing?

Reply:
(recording the same event multiple times should have ~no effect, anyway, but we're creating on first record so the CUDAEvent just doesn't bother)

Reply:
How about CUDASyncEvent?

Because if you dislike that then bonus is that we're passing a flag to the event that disables timing so it runs as quickly as possible for our sync case. This makes the intent of the event clear and means that recording to the same event multiple times is either an error or a silly thing to do.

Reply:
Well `Sync` doesn't really imply "record once". "an error or a silly thing to do" is pretty much what I thought when I saw this code, which is where the comment is coming from haha. I'd really rather make this logic explicit, or move the `.record` to happen only once, to avoid confusion in the future.

Reply:
How about one of: CUDAOnceEvent, OnetimeCUDAEvent, CUDAOnetimeEvent?

Reply:
Edit: @ngimel convinced me IdempotentCUDAEvent was a fine name.

Reply:
Alternatively we could `record` -> `idempotentRecord`/`recordOnce`

Reply:
(that will prevent a profusion of CUDAEvent classes if we ever want to expand their functionality, like you mention)